File: animations-java/src/main/java/dev/hogoshi/animations/animation/ContextAnimation.java
package dev.hogoshi.animations.animation;

import dev.hogoshi.animations.AnimationContext;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.RequiredArgsConstructor;

/**
 * Animation which stores as this class/object inside of context
 *
 * @author hogoshi
 */
@EqualsAndHashCode(callSuper = true)
@Data
@RequiredArgsConstructor
public class ContextAnimation extends AbstractAnimation {

    private final String name;
    private final AnimationContext context;

    /**
     * Use for animating animation inside AnimationContext's animates runnable
     */
    @Override
    public void setValue(double valueTo) {
        if (valueTo == getValue()) return;

        setDebug(getContext().isDebug());

        if (isAlive() && (valueTo == getFrom() || valueTo == getTo() || valueTo == getValue())) {
            if (isDebug()) System.out.println("Animating " + name + " cancelled due to valueTo equals fromValue");
            return;
        }

        setEasing(getContext().getEasing());
        setDuration(getContext().getDuration() * 1000);
        setStart(System.currentTimeMillis());
        setFrom(getValue());
        setTo(valueTo);

        if (isDebug())
            System.out.println(name + "#animate {\n    to value: " + getTo() + "\n    from value: " + getValue() + "\n    duration: " + getDuration() + "\n}");
    }

}

File: animations-kotlin/src/main/kotlin/dev/hogoshi/animations/animation/ContextAnimation.kt
package dev.hogoshi.animations.animation

import dev.hogoshi.animations.AnimationContext

/**
 * Animation which stores as this class/object inside of context
 * @author hogoshi
 */
class ContextAnimation(
    val name: String,
    private val context: AnimationContext
) : AbstractAnimation() {

    /**
     * Use for animating animation inside AnimationContext's animates runnable
     */
    fun animate(value: Double) {
        this.debug = context.debug
        if (isAlive() && (value == from || value == to || value == this.value)) {
            if (debug) println("Animating $name cancelled due to valueTo equals fromValue")
            return
        }

        this.easing = context.easing
        this.duration = context.duration * 1000
        this.start = System.currentTimeMillis()
        this.from = this.value
        this.to = value

        if (debug)
            println(
                """$name#animate {
                    to value: $to
                    from value: $from
                    duration: $duration
                }"""
            )
    }

}